### ISSUE TITLE
CustomFields Lost During Request Execution

### ISSUE THREAD

**Issue by SpaceCondor** [SAP/cloud-sdk#1370]
## Issue Description

When using the `setCustomField` method to set custom fields of elements of a collection, they are lost upon execution.

- Important information:

Given the attached generated code (created via the cli generator) the issue can be reproduced with the following code:

[odata.zip](https://github.com/SAP/cloud-sdk/files/11426574/odata.zip)

```
        HttpDestination httpDestination = DefaultHttpDestination.builder("http://localhost/").name("test").trustAllCertificates().build();

        DefaultDestinationLoader customLoader = new DefaultDestinationLoader().registerDestination(httpDestination);
        DestinationAccessor.appendDestinationLoader(customLoader);
        HttpClient httpClient = HttpClientAccessor.getHttpClient(httpDestination);

        Document sapDocument = new Document();

        sapDocument.setDocEntry(1);
        sapDocument.setDocumentPackages(new ArrayList<>());
        sapDocument.setCustomField("custom_field_test_document", "test_value_document");

        DocumentPackage documentPackage = new DocumentPackage();
        documentPackage.setCustomField("custom_field_test_package", "test_value_package");

        sapDocument.getDocumentPackages().add(documentPackage);

        DefaultBusinessOneService businessOneService = new DefaultBusinessOneService();
        UpdateRequestBuilder<Document> request = businessOneService.updateDeliveryNotes(sapDocument);

        request.execute(httpDestination);
```

- Expected outcome

The POST request should look similar to this:

```
{
  "DocEntry": 1,
  "DocumentPackages": [
    {
      "Type": null,
      "TotalWeight": null,
      "DocumentPackageItems": null,
      "Number": null,
      "Units": null,
      "custom_field_test_package": "test_value_package"
    }
  ],
  "custom_field_test_document": "test_value_document",
  "@odata.type": "#SAPB1.Document"
}
```

- Actual outcome

The custom field is lost

```
{
  "DocEntry": 1,
  "DocumentPackages": [
    {
      "Type": null,
      "TotalWeight": null,
      "DocumentPackageItems": null,
      "Number": null,
      "Units": null
    }
  ],
  "custom_field_test_document": "test_value_document",
  "@odata.type": "#SAPB1.Document"
}
```

- SAP Cloud SDK Version: 4.13.0

**Comment by MatKuhr** [comment]
Hi Conner, thanks for bringing this up! 

I think this is indeed a bug specific to complex properties in the default update strategy `modifyingEntity()`. We'll do a bit more testing and check if we can quickly provide a fix.

In the mean time, it looks like updates with `PUT` are not affected. Could you try this and confirm that with `replacingEntity()` the issue does not occur?

Also, I should mention that when using `modifyingEntity()` (which is the default) entities should first be retrieved, before they are used in an update. I.e. creating objects ( e.g. via `new Document()`) and then using them in update requests is not recommended. [This documentation](https://sap.github.io/cloud-sdk/docs/java/features/odata/v4-vdm-client#entity-update-strategies) gives further details.

**Comment by SpaceCondor** [comment]
Hey @MatKuhr,

Thank you for the info! After testing it looks like you are correct in that this does not occur when utilizing `replacingEntity()`

Also I understand about the suggestion to first retrieve the entity and then update, especially when considering usage of ETAGS. I will investigate the feasibility of this, but the nature of our usage requires many small updates to these entities which can slow down significantly if we retrieve before every update.

**Comment by MatKuhr** [comment]
Good to hear the workaround works as expected! We looked into the problem further and it looks like this may be fixable rather quickly. I'll let you know if/once a fix is available. 

------------

However, I do want to point out one other further limitation (that is currently not documented anywhere I think): If you have a **collection of complex properties**, and you add or remove items, this will not be detected and not be part of update requests. 

For example:
```java
TestEntity entity = TestEntity.builder().complexProperties(new ArrayList<>()).build();
entity.getComplexProperties().add(new TestComplex());
        
// this json will be empty
String json = service.updateTestEntity(entity).toRequest().getSerialisedEntity();
```

But you can enforce the property to be included using `.includingFields(..)`.

----------

Regarding the update strategy: Perhaps batch requests might solve the problem for you. You can group read requests together into a [batch operation](https://sap.github.io/cloud-sdk/docs/java/features/odata/v4-vdm-client#batch-requests) and receive all results at once. So batch is not only for sending create/update/delete requests, but read requests as well.

**Comment by SpaceCondor** [comment]
@MatKuhr Thank you for the update! And I appreciate the heads up about adding/removing elements of a collection, I will certainly keep an eye out for that. 

I will also look into the approach of using batch requests. However, I do not know if that will work. Just as an example, we have a field for "Shipment Status." This field gets updated very frequently but is just one piece of a very large entity. We really just need to send the request and know if it was successful, there's nothing to 'read' per-se.

**Comment by MatKuhr** [comment]
Okay, yeah, batch doesn't really help in case of frequent updates of the same entity, it only helps when reading / updating many entities. Then just be careful to really only set the fields you want to change when sending "new" objects via update. Be aware that in case of complex types the Cloud SDK currently sends all fields, not only changed ones. 

Also: we just [released version 4.14.0](https://sap.github.io/cloud-sdk/docs/java/release-notes#fixed-issues) where the issue has been fixed. So you should be able to use the default `modifyingEntity()` again.
